lib/cretonne/src/isa/intel/mod.rs

```diff
@@ -111,4 +117,145 @@ impl TargetIsa for Isa {
     }
+
+    fn prologue_epilogue(&self, func: &mut ir::Function) -> result::CtonResult {
+        let word_size = if self.flags().is_64bit() { 8 } else { 4 };
+        let csr_type = if self.flags().is_64bit() {
+            ir::types::I64
+        } else {
+            ir::types::I32
+        };
+        let csrs = abi::callee_saved_registers(&self.shared_flags);
+
+        let mut csr_stack_size = word_size; // Size of RBP to start with
+        for _reg in &csrs {
+            csr_stack_size += word_size;
+        }
+
+        let stack_offset = -(csr_stack_size as i32);
+        let slot = ir::StackSlotData {
+            kind: ir::StackSlotKind::IncomingArg,
+            size: csr_stack_size,
+            offset: stack_offset,
+        };
+        func.create_stack_slot(slot);
+
+        let total_stack_size = layout_stack(&mut func.stack_slots, word_size)?;
+        let local_stack_size = total_stack_size - csr_stack_size;
+
+        // Append frame pointer to function signature
+        let rbp_arg = ir::AbiParam::special_reg(
+            csr_type,
+            ir::ArgumentPurpose::FramePointer,
+            RU::rbp as RegUnit,
+        );
+        func.signature.params.push(rbp_arg);
+        func.signature.returns.push(rbp_arg);
+
+        for reg in &csrs {
+            let csr_arg = ir::AbiParam::special_reg(
+                csr_type,
+                ir::ArgumentPurpose::CalleeSaved,
+                *reg as RegUnit,
+            );
+            func.signature.params.push(csr_arg);
+            func.signature.returns.push(csr_arg);
+        }
+
+        // Append param to entry EBB
+        let entry_ebb = func.layout.entry_block().expect("missing entry block");
+        func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+        // Find our frame pointer parameter Value
+        let fp = func.special_param(ir::ArgumentPurpose::FramePointer)
+            .expect("missing frame pointer");
+
+        // Assign it a location
+        func.locations[fp] = ir::ValueLoc::Reg(RU::rbp as RegUnit);
+
+        let mut csr_vals = Vec::new();
+        for reg in &csrs {
+            // Append param to entry EBB
+            func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+            let csr_arg = func.dfg.ebb_params(entry_ebb).last().expect(
+                "no last argument",
+            );
+
+            // Assign it a location
+            func.locations[*csr_arg] = ir::ValueLoc::Reg(*reg as RegUnit);
+
+            // Remember it so we can push it momentarily
+            csr_vals.push(*csr_arg);
+        }
+
+
+        // Insert prologue
+        {
+            let mut pos = EncCursor::new(func, self).at_first_insertion_point(entry_ebb);
+            pos.ins().x86_push(fp);
+            pos.ins().copy_special(
+                RU::rsp as RegUnit,
+                RU::rbp as RegUnit,
+            );
+            if local_stack_size > 0 {
+                pos.ins().adjust_sp_imm(-(local_stack_size as i32));
```

Comment:
This should return a `CtonError::ImplLimitExceeded` if `local_stack_size` is greater than `i32::MAX as u32`, so that we don't overflow.

Reply:
The stack layout algorithm enforces a maximum 2 GB stack frame size for its own sanity. This should probably be better documented.